### ISSUE TITLE
LegendControl: System.InvalidOperationException: Object is currently in use elsewhere.

### ISSUE THREAD

**Issue by sindizzy** [MapWindow/MapWinGIS#70]
I have been struggling with this exception for many years and have not been able to find what causes it. When this exception occurs I get a big red X where the legend should be. Sometimes the user will close and reopen the app 3 or  times and then it will go away. Other times they have to reboot the machine. Other times it just goes back to normal on the first load.

Here is the exception details with pertinent info:

> See the end of this message for details on invoking 
> just-in-time (JIT) debugging instead of this dialog box.
> 
> ************** Exception Text **************
> System.InvalidOperationException: Object is currently in use elsewhere.
>    at System.Drawing.Graphics.Clear(Color color)
>    at LegendControl.Legend.DrawNextFrame()
>    at LegendControl.Legend.OnPaint(PaintEventArgs e)
>    at System.Windows.Forms.Control.PaintWithErrorHandling(PaintEventArgs e, Int16 layer)
>    at System.Windows.Forms.Control.WmPaint(Message& m)
>    at System.Windows.Forms.Control.WndProc(Message& m)
>    at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
>    at System.Windows.Forms.ContainerControl.WndProc(Message& m)
>    at System.Windows.Forms.UserControl.WndProc(Message& m)
>    at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
>   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
>    at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
> 
> 
> ************** Loaded Assemblies **************
> mscorlib
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.NET/Framework/v4.0.30319/mscorlib.dll
> ----------------------------------------
> Microsoft.VisualBasic
>     Assembly Version: 10.0.0.0
>     Win32 Version: 14.6.1099.0 built by: NETFXREL4STAGE
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/Microsoft.VisualBasic/v4.0_10.0.0.0__b03f5f7f11d50a3a/Microsoft.VisualBasic.dll
> ----------------------------------------
> System
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
> ----------------------------------------
> System.Core
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
> ----------------------------------------
> System.Windows.Forms
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
> ----------------------------------------
> System.Drawing
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
> ----------------------------------------
> System.Configuration
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.6.1099.0 built by: NETFXREL4STAGE
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
> ----------------------------------------
> System.Xml
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
> ----------------------------------------
> System.Runtime.Remoting
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Remoting/v4.0_4.0.0.0__b77a5c561934e089/System.Runtime.Remoting.dll
> ----------------------------------------
> Interop.MapWinGIS
>     Assembly Version: 4.8.0.0
>     Win32 Version: 4.8.0.0
>     CodeBase: file:///C:/Program%20Files%20(x86)/RouteMapper/Interop.MapWinGIS.DLL
> ----------------------------------------
> DotSpatial.Controls
>     Assembly Version: 2.0.0.349
>     Win32 Version: 2.0.0.349
>     CodeBase: file:///C:/Program%20Files%20(x86)/RouteMapper/DotSpatial.Controls.DLL
> ----------------------------------------
> MapWinInterfaces
>     Assembly Version: 4.8.6.42296
>     Win32 Version: 4.8.6.0
>     CodeBase: file:///C:/Program%20Files%20(x86)/RouteMapper/MapWinInterfaces.DLL
> ----------------------------------------
> AxInterop.MapWinGIS
>     Assembly Version: 4.8.0.0
>     Win32 Version: 4.8.0.0
>     CodeBase: file:///C:/Program%20Files%20(x86)/RouteMapper/AxInterop.MapWinGIS.DLL
> ----------------------------------------
> System.Data
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_32/System.Data/v4.0_4.0.0.0__b77a5c561934e089/System.Data.dll
> ----------------------------------------
> Accessibility
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.6.1099.0 built by: NETFXREL4STAGE
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/Accessibility/v4.0_4.0.0.0__b03f5f7f11d50a3a/Accessibility.dll
> ----------------------------------------
> System.Deployment
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Deployment/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Deployment.dll
> ----------------------------------------
> Microsoft.GeneratedCode
>     Assembly Version: 1.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
> ----------------------------------------
> System.ComponentModel.Composition
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.6.1099.0
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.ComponentModel.Composition/v4.0_4.0.0.0__b77a5c561934e089/System.ComponentModel.Composition.dll
> ----------------------------------------
> Anonymously Hosted DynamicMethods Assembly
>     Assembly Version: 0.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_32/mscorlib/v4.0_4.0.0.0__b77a5c561934e089/mscorlib.dll
> ----------------------------------------
> System.Data.DataSetExtensions
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.6.1099.0 built by: NETFXREL4STAGE
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Data.DataSetExtensions/v4.0_4.0.0.0__b77a5c561934e089/System.Data.DataSetExtensions.dll
> ----------------------------------------
> Microsoft.GeneratedCode
>     Assembly Version: 1.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
> ----------------------------------------
> System.Numerics
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.6.1099.0 built by: NETFXREL4STAGE
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Numerics/v4.0_4.0.0.0__b77a5c561934e089/System.Numerics.dll
> ----------------------------------------
> System.Transactions
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.7.2117.0 built by: NET47REL1LAST
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_32/System.Transactions/v4.0_4.0.0.0__b77a5c561934e089/System.Transactions.dll
> ----------------------------------------
> System.EnterpriseServices
>     Assembly Version: 4.0.0.0
>     Win32 Version: 4.6.1099.0 built by: NETFXREL4STAGE
>     CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_32/System.EnterpriseServices/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.EnterpriseServices.dll
> ----------------------------------------
> stdole
>     Assembly Version: 7.0.3300.0
>     Win32 Version: 7.00.9466
>     CodeBase: file:///C:/Windows/assembly/GAC/stdole/7.0.3300.0__b03f5f7f11d50a3a/stdole.dll
> ----------------------------------------
> 
> ************** JIT Debugging **************
> To enable just-in-time (JIT) debugging, the .config file for this
> application or computer (machine.config) must have the
> jitDebugging value set in the system.windows.forms section.
> The application must also be compiled with debugging
> enabled.
> 
> For example:
> 
> <configuration>
>     <system.windows.forms jitDebugging="true" />
> </configuration>
> 
> When JIT debugging is enabled, any unhandled exception
> will be sent to the JIT debugger registered on the computer
> rather than be handled by this dialog box.
>

**Comment by pmeems** [comment]
You are using a very old version of MapWinGIS. The problem you mention is one of the reasons we don't have a legend control anymore with the latest versions of MapWinGIS. We do have a small C# demo application (for MapWinGIS v4.9+) showing how to implement your own legend control.

It might be worthwhile to upgrade to MapWinGIS v4.9+
A lot is changed and has been improved. Depending on the amount of custom work it might be a big task. If you do want to upgrade you could also consider rebuilding your custom work as an MW5 plug-in. Then you'll have all the benefits of MapWindow5 and you'll only need to focus on your custom code.
Plug-in creation has been enormously simplified in MW5.

So in short: I cannot help you with this error. I haven't been working on v4.8 for years. You can, of course, download the source code and try to fix it yourself.

**Comment by sindizzy** [comment]
Thanks for the info. I have been trying to update MW on my projects for some time but its a slow process within my company. I'm also converting some to DotSpatial since I need the reprojection functions. I might take MW5 for a spin and see how much work it will be to upgrade.

**Comment by pmeems** [comment]
MapWinGIS uses GDAL/PROJ4 for its reprojection. It works very well.
Have a look at sf.Reproject(): http://www.mapwindow.org/documentation/mapwingis4.9/group__shapefile__geoprocessing.html#gaa248de52c40fadf103ce50d3c349d6da

**Comment by sindizzy** [comment]
Appreciate the suggestions. Several years ago when I tried the reprojection the user had to keep separate shapefiles for each projection. I see now a ReprojectInPlace that I assume reprojects the shapefile in memory without having to create new shapefiles on the disk. I may give this another shot.

**Comment by pmeems** [comment]
I'm closing this issue.
If needed you can repost at our [MapWinGIS forum](https://mapwindow.discourse.group/c/mapwingis)
